src/main/java/space/obminyashka/items_exchange/service/impl/AdvertisementServiceImpl.java

```diff
@@ -55,12 +55,17 @@ public Page<AdvertisementTitleDto> findAllThumbnails(Pageable pageable) {
 
     @Override
     @Cacheable
-    public List<AdvertisementTitleDto> findRandom12Thumbnails() {
+    public List<AdvertisementTitleDto> findRandomNThumbnails(int amount, UUID advertisementId, Long subcategoryId) {
         final var totalRecordsSize = count();
-        final var resultsQuantity = 12;
+        final var resultsQuantity = amount;
         final var bound = (int) (totalRecordsSize / resultsQuantity);
         final var randomPage = bound > 0 ? random.nextInt(bound) : 0;
-        return findAllThumbnails(PageRequest.of(randomPage, resultsQuantity)).getContent();
+        return ((advertisementId != null && subcategoryId != null) ? findAllThumbnails(PageRequest.of(randomPage, resultsQuantity), advertisementId, subcategoryId) : findAllThumbnails(PageRequest.of(randomPage, resultsQuantity))).getContent();
```

Comment:
The not null check has to be removed and only 1 unified request method should exist that would search all non-null parameters.